truffle/src/main/java/org/jruby/truffle/core/klass/ClassNodes.java

```diff
@@ -330,8 +333,30 @@ public SuperClassNode(RubyContext context, SourceSection sourceSection) {
             super(context, sourceSection);
         }
 
+        @Specialization(guards = { "rubyClass == cachedRubyCLass", "cachedSuperclass != null" }, limit = "getCacheLimit()")
+        public Object getSuperClass(DynamicObject rubyClass,
+                                    @Cached("rubyClass") DynamicObject cachedRubyCLass,
+                                    @Cached("fastLookUp(cachedRubyCLass)") DynamicObject cachedSuperclass) {
+            // caches only initialized classes, just allocated will go through slow look up
+            return cachedSuperclass;
+        }
+
         @Specialization
-        public Object getSuperClass(DynamicObject rubyClass) {
+        DynamicObject getGetSuperClass(DynamicObject rubyClass) {
```

Comment:
I am not sure what's our usual approach to this. Leave few cache instances using generic specialisation for rest or using only generic specialisation when cache is not big enough? Here it would probably make more sense to add contains as Kevin suggests.